api/src/main/java/org/pmiops/workbench/reporting/ReportingVerificationServiceImpl.java

```diff
@@ -0,0 +1,154 @@
+    verifyStopwatch.start();
+
+    final ReportingUploadDetails result =
+        new ReportingUploadDetails()
+            .snapshotTimestamp(snapshot.getCaptureTimestamp())
+            .projectId(getProjectId())
+            .dataset(getDataset())
+            .uploads(
+                ImmutableList.of(
+                    checkUpload(
+                        UserColumnValueExtractor.TABLE_NAME,
+                        snapshot.getCaptureTimestamp(),
+                        snapshot.getUsers().size()),
+                    checkUpload(
+                        WorkspaceColumnValueExtractor.TABLE_NAME,
+                        snapshot.getCaptureTimestamp(),
+                        snapshot.getWorkspaces().size()),
+                    checkUpload(
+                        CohortColumnValueExtractor.TABLE_NAME,
+                        snapshot.getCaptureTimestamp(),
+                        snapshot.getCohorts().size()),
+                    checkUpload(
+                        InstitutionColumnValueExtractor.TABLE_NAME,
+                        snapshot.getCaptureTimestamp(),
+                        snapshot.getInstitutions().size())));
+    verifyStopwatch.stop();
+    logger.info(LogFormatters.duration("Verification queries", verifyStopwatch.elapsed()));
+    return result;
+  }
+
+  @Override
+  public ReportingUploadDetails verifyAndLog(ReportingSnapshot reportingSnapshot) {
+    // check each table. Note that for streaming inputs, not all rows may be immediately available.
+    final ReportingUploadDetails uploadDetails = checkUploads(reportingSnapshot);
+    final StringBuilder sb =
+        new StringBuilder(
+            String.format("Verifying Snapshot %d:\n", reportingSnapshot.getCaptureTimestamp()));
+    sb.append("Table\tUploaded\tExpected\tDifference(%)\n");
+    Level detailsLogLevel = Level.INFO;
+    for (final ReportingUploadResult result : uploadDetails.getUploads()) {
+      final long delta = result.getDestinationRowCount() - result.getSourceRowCount();
+      final String relativeDelta;
+      if (result.getSourceRowCount() == 0) {
+        relativeDelta = "";
+      } else {
+        relativeDelta = String.format(" (%.3f%%)", 100.0 * delta / result.getSourceRowCount());
+      }
+      sb.append(
+          String.format(
+              "%s\t%d\t%d\t%d%s\n",
+              result.getTableName(),
+              result.getSourceRowCount(),
+              result.getDestinationRowCount(),
+              delta,
+              relativeDelta));
+      if (!result.getDestinationRowCount().equals(result.getSourceRowCount())) {
+        detailsLogLevel = Level.WARNING;
+      }
+    }
+    logger.log(detailsLogLevel, sb.toString());
+    return uploadDetails;
+  }
+
+  public String getDataset() {
+    return workbenchConfigProvider.get().reporting.dataset;
+  }
+
+  public String getProjectId() {
+    return workbenchConfigProvider.get().server.projectId;
+  }
+
+  private ReportingUploadResult checkUpload(
```

Comment:
It doesn't look like this is checking anything, would something like `buildReportingUploadResult` be a better name?